### PR TITLE
HYDRATOR-378 add checkpointing to data streams app

### PR DESCRIPTION
**Pull request by albertshau**
This involves adding a FileSet that will be used to write
checkpoints for all pipelines. The change also requires some
refactoring since all driver functions now need to be
serializable, due to how Spark streaming checkpoints work.
As such, moving all the anonymous classes in the DStreamCollection
class into their own classes. Also had to tweak functions are
serialized/deserialized so that the plugin is instantiated and
initialized every time instead of just once in the driver. This
is so that macro evaluation will be run each time a pipeline is
run, instead of just the very first time the pipeline is run and
then serialized.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT4919

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java [hunk 1]
```diff
@@ -92,6 +104,46 @@ public void initialize() throws Exception {
       sparkConf.setMaster(String.format("local[%d]", numSources + 1));
     }
     context.setSparkConf(sparkConf);
+
+    if (!checkpointsDisabled) {
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java:113]: Is there a reason why it has to checkpointDisabled ? if we have checkpointEnabled then we can get rid of this ! thingy all over.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java [hunk 2]
```diff
@@ -100,26 +73,78 @@ public void run(JavaSparkExecutionContext sec) throws Exception {
...
+    final PipelinePhase pipelinePhase = phaseBuilder.build();
 
-    sparkContext = new JavaSparkContext();
-    streamingContext = new JavaStreamingContext(sparkContext,
-                                                Durations.milliseconds(pipelineSpec.getBatchIntervalMillis()));
-    // TODO: figure out how to get partitions to use for aggregators and joiners.
-    // Seems like they should be set at configure time instead of runtime? but that requires an API change.
-    runPipeline(pipelinePhase, StreamingSource.PLUGIN_TYPE, sec, new HashMap<String, Integer>());
+    boolean checkpointsDisabled = Boolean.parseBoolean(
+      sec.getSpecification().getProperty("cask.hydrator.checkpoint.disabled"));
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java]: I  remember when we  change names from continuuity to cask there was a discussion not to put company name in classsnames, properties etc.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java [hunk 3]
```diff
@@ -100,26 +73,78 @@ public void run(JavaSparkExecutionContext sec) throws Exception {
...
     }
   }
+
+  private JavaStreamingContext runWithoutCheckpointing(final DataStreamsPipelineSpec pipelineSpec,
+                                                       final PipelinePhase pipelinePhase,
+                                                       final JavaSparkExecutionContext sec) throws Exception {
+    JavaStreamingContext jssc = new JavaStreamingContext(
+      new JavaSparkContext(), Durations.milliseconds(pipelineSpec.getBatchIntervalMillis()));
+    SparkStreamingPipelineRunner runner = new SparkStreamingPipelineRunner(sec, jssc, true);
+    // TODO: figure out how to get partitions to use for aggregators and joiners.
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java]: Should we add  a jira to track this ?

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java [hunk 4]
```diff
@@ -51,6 +52,7 @@
 import org.junit.Test;
 
 import java.util.ArrayList;
+import java.util.HashMap;
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java:55]: APP_ARTIFACT_ID and APP_ARTIFACT can be private now.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/DataStreamsConfig.java [hunk 5]
```diff
@@ -38,12 +39,14 @@ private DataStreamsConfig(Set<ETLStage> stages,
                             Resources driverResources,
                             boolean stageLoggingEnabled,
                             String batchInterval,
-                            boolean isUnitTest) {
+                            boolean isUnitTest,
+                            boolean disableCheckpoints) {
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/co/cask/cdap/etl/proto/v2/DataStreamsConfig.java]: Doesn't the equals, hashcode and toString needs to be updated ?

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/SparkPairCollection.java [hunk 6]
```diff
@@ -38,10 +38,6 @@
 
   <T> SparkPairCollection<K, T> mapValues(Function<V, T> function);
 
-  SparkPairCollection<K, Iterable<V>> groupByKey();
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/SparkPairCollection.java:41]: Why remove these ?

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java [hunk 7]
```diff
@@ -55,10 +48,18 @@
...
-public abstract class SparkPipelineDriver {
+public abstract class SparkPipelineRunner {
 
-  protected abstract SparkCollection<Object> getSource(String stageName,
-                                                       PluginFunctionContext pluginFunctionContext) throws Exception;
+  protected abstract SparkCollection<Object> getSource(StageInfo stageInfo) throws Exception;
+
+
+  protected abstract SparkPairCollection<Object, Object> addJoinKey(
+    StageInfo stageInfo, String inputStageName,
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java:57]: Should we pull some paras up. I dont think we format like this

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/joiner/MockJoiner.java [hunk 8]
```diff
@@ -16,6 +16,7 @@
 
 package co.cask.cdap.etl.mock.batch.joiner;
 
+import co.cask.cdap.api.annotation.Macro;
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/joiner/MockJoiner.java]: This is unused right ?

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/joiner/DupeFlagger.java [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+      outputBuilder.set(field.getName(), record.get(field.getName()));
+    }
+    return outputBuilder.build();
+  }
+
+  /**
+   * Config for except plugin
+   */
+  public static class Config extends PluginConfig {
+    @Macro
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/joiner/DupeFlagger.java:130]: Should we have descriptions for these fields ?

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java (hunk 1 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java:113]: because the default behavior is to enable checkpoints. The default value of a boolean is false, so if the property was 'checkpointEnabled', the absence of the property would imply checkpointEnabled=false.

that being said, it only matters in the app config. I can flip it here if its really a concern, but thought it would be better to be consistent with whats in the config.

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/streaming/function/DynamicAggregatorAggregate.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  private final DynamicDriverContext dynamicDriverContext;
+  private transient AggregatorAggregateFunction<GROUP_KEY, GROUP_VAL, OUT> function;
+
+  public DynamicAggregatorAggregate(DynamicDriverContext dynamicDriverContext) {
+    this.dynamicDriverContext = dynamicDriverContext;
+  }
+
+  @Override
+  public JavaRDD<OUT> call(JavaPairRDD<GROUP_KEY, Iterable<GROUP_VAL>> input, Time batchTime) throws Exception {
+    if (function == null) {
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/streaming/function/DynamicAggregatorAggregate.java:45]: When can function be not null ?

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/streaming/function/DynamicJoinMerge.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+ */
+public class DynamicJoinMerge<JOIN_KEY, INPUT_RECORD, OUT>
+  implements Function2<JavaPairRDD<JOIN_KEY, List<JoinElement<INPUT_RECORD>>>, Time, JavaRDD<OUT>> {
+  private final DynamicDriverContext dynamicDriverContext;
+  private transient JoinMergeFunction<JOIN_KEY, INPUT_RECORD, OUT> function;
+
+  public DynamicJoinMerge(DynamicDriverContext dynamicDriverContext) {
+    this.dynamicDriverContext = dynamicDriverContext;
+  }
+
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/streaming/function/DynamicJoinMerge.java]: extra new line

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java (hunk 3 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java]: there is a jira HYDRATOR-265 already that I will add to the comment.

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/SparkPairCollection.java (hunk 6 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/SparkPairCollection.java:41]: they're not used anymore

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java (hunk 7 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java:57]: I can't, the last argument will break checkstyle because the line will be too long.

**Review by rohitsinha54 — COMMENTED** [review]
Took one pass. Some minor comments. But please ask @chtyim to look at it too.

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/streaming/function/DynamicAggregatorAggregate.java (hunk 10 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/streaming/function/DynamicAggregatorAggregate.java]: the first time it is called. This function is serialized by spark and stored in the checkpoint.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java (hunk 3 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java]: opened HYDRATOR-1030, will put it in the comment

File: cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/streaming/function/DynamicAggregatorAggregate.java (hunk 10 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/streaming/function/DynamicAggregatorAggregate.java:45]: this is called for each rdd in a dstream, so anything after the first rdd.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java [hunk 12]
```diff
@@ -79,19 +96,59 @@ public void initialize() throws Exception {
...
+      // Each pipeline has its own checkpoint directory within the checkpoint fileset.
+      // Ideally, when a pipeline is deleted, we would be able to delete that checkpoint directory.
+      // This is because we don't want another pipeline created with the same name to pick up the old checkpoint.
+      // Since CDAP has no way to run application logic on deletion, we instead generate a unique pipeline id
+      // and use that as the checkpoint directory as a subdirectory inside the pipeline name directory.
+      // On start, we check for any other pipeline ids for that pipeline name, and delete them if they exist.
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java:119]: What will happen if we put application version into the picture? I know that hydrator is not going to expose it yet in 4.0, but it'll be good to plan for it.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java [hunk 13]
```diff
@@ -79,19 +96,59 @@ public void initialize() throws Exception {
...
+      FileSet checkpointFileSet = context.getDataset(DataStreamsApp.CHECKPOINT_FILESET);
+      String pipelineName = context.getApplicationSpecification().getName();
+      String pipelineId = context.getSpecification().getProperty(PIPELINE_UUID);
+      Location pipelineCheckpointBase = checkpointFileSet.getBaseLocation().append(pipelineName);
+      Location pipelineCheckpointDir = pipelineCheckpointBase.append(pipelineId);
+
+      if (!pipelineCheckpointBase.exists()) {
+        if (!pipelineCheckpointBase.mkdirs()) {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java]: Technically there can be race when two pipeline tries to create the same base directory. A better way is to do:

``` java
if (!location.isDirectory() && !location.mkdirs() && !location.isDirectory()) {
  throws ...
}
```

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java [hunk 14]
```diff
@@ -79,19 +96,59 @@ public void initialize() throws Exception {
...
+          throw new IOException(
+            String.format("Unable to create checkpoint base directory '%s' for the pipeline.", pipelineCheckpointBase));
+        }
+      }
+
+      try {
+        for (Location child : pipelineCheckpointBase.list()) {
+          if (!child.equals(pipelineCheckpointDir)) {
+            if (!child.delete(true)) {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java]: You can merge the two `if` with `&&`

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java [hunk 15]
```diff
@@ -79,19 +96,59 @@ public void initialize() throws Exception {
...
+              LOG.warn("Unable to delete checkpoint directory {} from an old pipeline.", child);
+            }
+          }
+        }
+      } catch (Exception e) {
+        LOG.warn("Unable to clean up old checkpoint directories from old pipelines.", e);
+      }
+
+      if (!pipelineCheckpointDir.exists()) {
+        if (!pipelineCheckpointDir.mkdirs()) {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java]: Same race condition as above.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java [hunk 16]
```diff
@@ -100,26 +73,78 @@ public void run(JavaSparkExecutionContext sec) throws Exception {
...
+        checkpointBaseRef.set(checkpointFileSet.getBaseLocation());
+      }
+    });
+    Location pipelineCheckpointDir = checkpointBaseRef.get().append(pipelineName).append(pipelineId);
+    final String checkpointDir = pipelineCheckpointDir.toURI().toString();
+
+    JavaStreamingContextFactory contextFactory = new JavaStreamingContextFactory() {
+      @Override
+      public JavaStreamingContext create() {
+        JavaStreamingContext jssc = new JavaStreamingContext(
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java:124]: This is basically the same as the `runWithoutCheckpointing`, except the call to `jssc.checkpoint`, right? Better have a method `run(@Nullable Location checkoutLocation)`.

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java [hunk 17]
```diff
@@ -80,7 +82,7 @@ public static void setupTest() throws Exception {
   }
 
   @Test
-  public void testTransformCompute() throws Exception {
+  public void testTransformComputeWithMacros() throws Exception {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java:85]: Is this change related to addition of checkpoint?

Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-data-streams/src/test/java/co/cask/cdap/datastreams/DataStreamsTest.java:85]: Yes it is meant to make sure that macros are substituted correctly when the pipeline is loaded from a checkpoint (my first implementation did not do this correctly).

File: cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java (hunk 12 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/DataStreamsSparkLauncher.java:119]: That's a good question. One option would be to rename this property from 'pipelineid' to 'checkpointdir', and to make checkpointdir a property in the application config. If one is not given, the app will generate one as a UUID during configure time. Otherwise, it will use whatever the user supplies. 

If a user wants to add a new version of the pipeline, they can look up the checkpointdir from the previous version since it is just an application property, and set it in the application config.

**Comment by chtyim** [comment]
LGTM

**Comment by albertshau** [comment]
squashed changes, will merge when build is green.
